luna/gateware/interface/psram.py

```diff
@@ -405,3 +405,521 @@ def elaborate(self, platform):
+
+                # Otherwise, react with either a short period of latency
+                # or a longer one, depending on what the RAM requested via
+                # RWDS.
+                with m.Else():
+                    m.next = "HANDLE_LATENCY"
+
+                    # FIXME: our HyperRAM part has a fixed latency, but we could need to detect 
+                    # different variants from the configuration register in the future.
+                    with m.If(extra_latency | 1):
+                        m.d.sync += latency_clocks_remaining.eq(self.HIGH_LATENCY_CLOCKS)
+                    with m.Else():
+                        m.d.sync += latency_clocks_remaining.eq(self.LOW_LATENCY_CLOCKS)
+
+
+            # HANDLE_LATENCY -- applies clock cycles until our latency period is over.
+            with m.State('HANDLE_LATENCY'):
+                m.d.sync += latency_clocks_remaining.eq(latency_clocks_remaining - 1)
+
+                with m.If(latency_clocks_remaining == 0):
+                    with m.If(is_read):
+                        m.next = 'READ_DATA'
+                    with m.Else():
+                        m.next = 'WRITE_DATA'
+
+
+            # READ_DATA -- reads words from the PSRAM
+            with m.State('READ_DATA'):
+                m.d.sync += self.phy.read.eq(0b11)
+
+                datavalid_delay = Signal()
+                m.d.sync += datavalid_delay.eq(self.phy.datavalid)
+
+                with m.If(self.phy.datavalid):
+                    m.d.comb += [
+                        self.read_data     .eq(self.phy.dq.i),
+                        self.read_ready    .eq(1),
+                    ]
+
+                    # If our controller is done with the transaction, end it.
+                    with m.If(self.final_word):
+                        m.d.sync += self.phy.clk_en.eq(0),
+                        m.next = 'RECOVERY'
+
+            # WRITE_DATA -- write a word to the PSRAM
+            with m.State("WRITE_DATA"):
+                m.d.sync += [
+                    self.phy.dq.o    .eq(self.write_data),
+                    self.phy.dq.e    .eq(1),
+                    self.phy.rwds.e  .eq(~is_register),
+                    self.phy.rwds.o  .eq(0),
+                ]
+                m.d.comb += self.write_ready.eq(1),
+
+                # If we just finished a register write, we're done -- there's no need for recovery.
+                with m.If(is_register):
+                    m.next = 'IDLE'
+
+                with m.Elif(self.final_word):
+                    m.next = 'RECOVERY'
+
+
+            # RECOVERY state: wait for the required period of time before a new transaction
+            with m.State('RECOVERY'):
+                m.d.sync += self.phy.clk_en .eq(0)
+
+                # TODO: implement recovery
+                m.next = 'IDLE'
+
+
+
+        return m
+
+
+class HyperRAMDQSPHY(Elaboratable):
+    """ Gateware interface to HyperRAM series self-refreshing DRAM chips.
```

Comment:
Could you update the docstrings of this and `HyperRAMDQSInterface` to add some DQS reference?